datafusion/functions-nested/src/string.rs

```diff
@@ -524,63 +687,90 @@ pub fn string_to_array_inner<T: OffsetSizeTrait>(args: &[ArrayRef]) -> Result<Ar
+                        } else {
+                            list_builder.values().append_value(string);
+                        }
+                        list_builder.append(true);
+                    }
+                    (Some(string), Some(delimiter)) => {
+                        string.split(delimiter).for_each(|s| {
+                            if Some(s) == null_value {
                                 list_builder.values().append_null();
                             } else {
-                                list_builder.values().append_value(string);
+                                list_builder.values().append_value(s);
                             }
-                            list_builder.append(true);
-                        }
-                        (Some(string), Some(delimiter)) => {
-                            string.split(delimiter).for_each(|s| {
-                                if Some(s) == null_value {
-                                    list_builder.values().append_null();
-                                } else {
-                                    list_builder.values().append_value(s);
-                                }
-                            });
-                            list_builder.append(true);
-                        }
-                        (Some(string), None) => {
-                            string.chars().map(|c| c.to_string()).for_each(|c| {
-                                if Some(c.as_str()) == null_value {
-                                    list_builder.values().append_null();
-                                } else {
-                                    list_builder.values().append_value(c);
-                                }
-                            });
-                            list_builder.append(true);
-                        }
-                        _ => list_builder.append(false), // null value
+                        });
+                        list_builder.append(true);
                     }
-                });
-        }
-        _ => {
-            return exec_err!(
-                "Expect string_to_array function to take two or three parameters"
-            )
-        }
-    }
+                    (Some(string), None) => {
+                        string.chars().map(|c| c.to_string()).for_each(|c| {
+                            if Some(c.as_str()) == null_value {
+                                list_builder.values().append_null();
+                            } else {
+                                list_builder.values().append_value(c.as_str());
+                            }
+                        });
+                        list_builder.append(true);
+                    }
+                    _ => list_builder.append(false), // null value
+                }
+            }),
+    };
 
     let list_array = list_builder.finish();
     Ok(Arc::new(list_array) as ArrayRef)
 }
+
+trait StringArrayBuilderType: ArrayBuilder {
```

Comment:
this is an interesting idea -- maybe something worth looking into making more visible somehow / reusing to reduce duplication

Reply:
I really wish Rust had enum variants as types - with that I could really reduce a lot of this code. One of the reasons this took so long is that I was looking for a clean and fast way to handle the string types in a super nice way. I failed unfortunately - what I came come up with similar to https://github.com/Omega359/arrow-datafusion/blob/38d9bdde69381b4a7516346f862310700b6eb96e/datafusion/functions/src/utils.rs#L192 but that just has more overhead than is acceptable :(